architecture/2023-12-asyncworker-progressInfo.md

```diff
@@ -0,0 +1,130 @@
+
+The proposal here is to add a mechanism so that the running operation can send out progress information to the async worker. As a result, when the deployment operation eventually times out, along with the timed out message, the user will also be able to see the operation progress so far in the error message which could help in troubleshooting the deployment failure.
+
+## Terms and definitions
+
+AsyncWorker - The current architecture which assigns workers to process queued long-running jobs asynchronously.
+
+## Objectives
+
+> **Issue Reference:** https://github.com/radius-project/radius/issues/6835
+
+### Goals
+- Enable reporting of operation progress in async worker
+
+### Non-Goals
+- Modify reporting of operation progress in CLI
+
+### User scenarios (optional)
+
+#### User story 1
+
+The readiness probe for the application pod fails. This is a non-terminal failure condition. The current code keeps waiting for the deployment to complete and eventually the deployment times out. The error message to the user is "Deployment timed out after xxx seconds".
+
+If we implement https://github.com/radius-project/radius/issues/6284 and look at pod events, we can detect the readiness probe failure. However, we still need a way to report this information to the user.
+
+#### User story 2
+
+## Design
+
+In the current design, the following is used to track operation status:-
+
+```
+// AsyncOperationStatus represents an OperationStatus resource.
+type AsyncOperationStatus struct {
+	// Id represents the async operation id.
+	ID string `json:"id,omitempty"`
+
+	// Name represents the async operation name and is usually set to the async operation id.
+	Name string `json:"name,omitempty"`
+
+	// Status represents the provisioning state of the resource.
+	Status ProvisioningState `json:"status,omitempty"`
+
+	// StartTime represents the async operation start time.
+	StartTime time.Time `json:"startTime,omitempty"`
+
+	// EndTime represents the async operation end time.
+	EndTime *time.Time `json:"endTime,omitempty"`
+
+	// Error represents the error occurred during provisioning.
+	Error *ErrorDetails `json:"error,omitempty"`
+}
+```
+
+### Design details
+
+We will add the following field to AsyncOperationStatus to report the operation progress:-
+
+```
+// AsyncOperationStatus represents an OperationStatus resource.
+type AsyncOperationStatus struct {
+  ....
+
+  // ProgressEvents represents the progress of the async operation.
+  // The events are represented as a map of key-value pairs where the key is 
+  // the event type such as Info, Error and the value is the event message.
+  ProgressEvents map[string]string `json:"progressEvents,omitempty"`
```

Comment:
How does client gets this ProgressEvents? What's the protocol to get this error as client perspective? In ARM, [operationstatuses api](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#azure-asyncoperation-resource-format) is the interface to get the status of async operation.

Therefore, I do not think we need to add new property for process event.  The existing `errors.details` array property is enough to capture all in-progress error event.  See [this error format](https://github.com/Azure/azure-resource-manager-rpc/blob/8231d7df51aec87e6ccb5bcc04478136758b0a4c/v1.0/common-api-details.md#error-response-content)

```golang
	// Error represents the error occurred during provisioning.
	Error *ErrorDetails `json:"error,omitempty"`
```

In terms of implementation, we need to pass additional error channel to async worker controller to allow the controller to update the live error. then consumer side can keep adding the error into `error.Details` array

Reply:
Does that work in the case of success? My concern is that clients will be confused by the presence of the `error` field for a success case.

Reply:
<img width="1043" alt="image" src="https://github.com/radius-project/design-notes/assets/11863108/c03481c5-c6db-42ba-8751-f3da58c5ad00">

Ah make sense. Then, I think we can consider using `operations` array property. What do you think ?

Reply:
Where are we going to get the ProgressEvents from? Do we need a formatter for this so that the end user can see it properly?

Reply:
I wasn't aware there was an existing API for this. If there is an existing API, we should do another iteration with the proposed usage.